opam, aspcud: init packages for external solver

hakuch opened the pull request:
###### Motivation for this change

The [opam](https://opam.ocaml.org/) package manager depends on external solvers to solve packaging constraints.

While in theory opam can attempt to satisfy constraints with its internal solver, in practice, even having a small handful of packages installed can lead to time-outs.

Aspcud is the best supported and most popular external solver for opam, but, unlike opam, was not packaged with nixpkgs.

Aspcud depends on two complementary package: clasp, and gringo.

The patch to gringo was necessary to get it to compile (the code used `std::pow` without including `<cmath>`). Likewise, aspcud needs to be compiled with an older version of Boost.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [X] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

hakuch committed:
opam, aspcud: init packages for external solver

The opam package manager relies on external solvers to determine package
management decisions it makes related to upgrades, new installations,
etc.

While, strictly speaking, an external solver is optional, aspcud is
highly recommended in documentation. Furthermore, even having a
relatively small number of packages installed quickly causes the limits
of the interal solver to be reached (before it times out).

Aspcud itself depends on two programs from the same suite: gringo, and
clasp.

On Darwin, Boost 1.55 (and thus Gringo) do not build, so we only support
Aspcud on non-Darwin platforms.

hakuch commented:
I'm not sure why the CI failures are happening, as everything works fine on my local machine. Does anyone have any tips?

joachifm commented on pkgs/tools/misc/aspcud/default.nix:
```diff
@@ -0,0 +1,43 @@
+  meta = with stdenv.lib; {
+    description = "Aspcud is a solver for package problems in CUDF format using ASP";
+    homepage = http://potasssco.sourceforge.net/;
+    platform = platform.all;
```
This should be `platform = platforms.all` (missing s).

joachifm commented on pkgs/tools/misc/aspcud/default.nix:
```diff
@@ -0,0 +1,43 @@
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Aspcud is a solver for package problems in CUDF format using ASP";
```
Please remove the package name from `meta.description`

joachifm commented on pkgs/tools/misc/aspcud/default.nix:
```diff
@@ -0,0 +1,43 @@
+    platform = platform.all;
+    license = licenses.gpl3;
+  };
+}
```
Please add a trailing newline

joachifm commented:
@hakuch in general travis CI is only to catch low hanging fruit, it is not a requirement for merging. In this case, it does reveal a real problem, however. The reason you're not seeing this on your end is most likely that you're "only" doing a build test, which won't evaluate the meta attributes. To force evaluation of the meta attributes, you can do something like `nix-instantiate --eval --strict -A foo.meta`.

joachifm commented:
@henrytill as the opam maintainer, what do you think?

henrytill commented:
With the suggested changes, LGTM.  

I can definitely confirm that `opam`'s internal solver will crap out on trivial updates, so I agree with @hakuch that depending on `aspcud` is a good default.

hakuch commented:
It seems to have trouble on OS X, but we're getting there. Thanks for the feedback! Much appreciated.

hakuch commented:
@henrytill Looking at the failure on OS X, it's not clear to me what's failing. Any ideas?

henrytill commented:
@hakuch: I'll give it a look on my OS X machine.

henrytill commented:
@hakuch: `gringo` fix for clang stdenvs here: https://github.com/henrytill/nixpkgs/commit/df65b281b3605f392d8bb35d449eacec021e13bd. Unfortunately, GitHub won't let me open a PR against your fork.

However, `aspcud` seems to depend on `boost155` which is currently broken on OS X (https://github.com/NixOS/nixpkgs/issues/17759), bumping to `boost` (currently, `boost-1.60.0`) causes the `aspcud` build to fail.

So it looks like we have a few choices:

1. Don't depend on `aspcud` on OS X.  While, as mentioned above, the internal solver doesn't seem to be strong enough for everyday usage at this point, it is possible to work around this using a remote solver like http://cudf-solvers.irill.org/index.html.
2. Don't make `opam` depend on `aspcud` on other platforms, since, technically speaking, it is an optional (but _recommended_) runtime dependency.
3. Remove OS X support for `opam` until the above issue with `boost155` is fixed.  This is probably not a good idea, as some of our OCaml modules use `opam-installer --script` in their build steps.

I'm in favor of the first choice.

cc OCamlers @vbgl @wizeman @FlorentBecker @ericbmerritt ... any opinions?

vbgl commented:
Indeed, neither `gringo` nor `boost155` will build on darwin. Option 1 looks the best to me: you may add a `aspcudSuppor` option to the derivation that defaults to `!stdenv.isDarwin`.

hakuch commented:
The latest version of this PR makes `aspcud` and (some of) its dependent packages Linux-only, and then makes @vbgl's recommendation for an optional dependency on `aspcud` in `opam`.

I'll check in a bit later, and please let me know if I misinterpreted the helpful suggestions.

vbgl commented:
`clasp` does not build on NixOS: the configure script calls `/bin/bash`. This can be fixed by adding a `preConfigure = "patchShebangs ./configure.sh"` attribute to the `clasp` derivation.

@hakuch would you like to list yourself as maintainer of these packages?

The license of `aspcud` should be `gpl3Plus` rather than `gpl3` (this is what `./aspcud -v` tells me).

hakuch commented:
Fixed licenses and updated the `maintainer` field (I assume I was supposed to add an entry in the maintainer file. Please let me know if I shouldn't have done this).

hakuch commented:
Hm... Unless I'm mistaken, the `ocaml-x509` package is failing to build on OS X. The log from the CI failure is [here](https://s3.amazonaws.com/archive.travis-ci.org/jobs/156694334/log.txt).

Sorry to bother you again, @henrytill, but any ideas?

vbgl commented:
I cannot reproduce the `x509` error locally. Maybe it is just a travis quirk…

Also, `clasp` and `aspcud` should have a `meta.platforms` attribute rather than a `meta.platform` one (notice the missing final -s).

hakuch commented:
I fixed the `meta.platform` attributes. Thanks!
